File: php/src/LombokAlgoritma/Graph/FloydWarshall.php
<?php

// LombokAlgoritma — Floyd–Warshall all-pairs shortest paths (SPEC §9.6)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

declare(strict_types=1);

namespace LombokAlgoritma\Graph;

/** Floyd–Warshall, loop order k → i → j with a strict `<` update. O(V³). */
final class FloydWarshall
{
    /**
     * All-pairs distances (INF when unreachable); parallel edges keep the minimum weight.
     *
     * @return list<list<int|float>>
     */
    public static function allPairs(Graph $g): array
    {
        $n = $g->nodes;
        $dist = [];
        for ($i = 0; $i < $n; $i++) {
            $row = array_fill(0, $n, INF);
            $row[$i] = 0;
            $dist[] = $row;
        }
        foreach ($g->edges as $e) {
            $dist[$e->from][$e->to] = min($dist[$e->from][$e->to], $e->weight);
        }
        for ($k = 0; $k < $n; $k++) {
            for ($i = 0; $i < $n; $i++) {
                $dik = $dist[$i][$k];
                if ($dik === INF) {
                    continue;
                }
                for ($j = 0; $j < $n; $j++) {
                    $cand = $dik + $dist[$k][$j];
                    if ($cand < $dist[$i][$j]) {
                        $dist[$i][$j] = $cand;
                    }
                }
            }
        }
        return $dist;
    }
}

File: python/lombokalgoritma/graph/floyd_warshall.py
# LombokAlgoritma — Floyd–Warshall all-pairs shortest paths (SPEC §9.6)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

import math

from .types import Graph, validate_graph


def floyd_warshall(g: Graph) -> list[list[float]]:
    """All-pairs distances (``inf`` when unreachable); parallel edges keep the minimum."""
    validate_graph(g)
    n = int(g.nodes)
    dist: list[list[float]] = [[0 if i == j else math.inf for j in range(n)] for i in range(n)]
    for u, v, w in g.edges:
        dist[u][v] = min(dist[u][v], w)
    for k in range(n):
        dk = dist[k]
        for i in range(n):
            di = dist[i]
            dik = di[k]
            if dik == math.inf:
                continue
            for j in range(n):
                cand = dik + dk[j]
                if cand < di[j]:
                    di[j] = cand
    return dist
